### <h1 align="center"> Competition II</h1>
<h3 align="center"> IST 5520: Data Methodologies in Python </h3>
<h3 align="center"> By: SHALINI RALLAPALLI VENKATA </h3>

In [257]:
# Import modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
import statsmodels.api as sm
from patsy import dmatrices
from sklearn import model_selection
from sklearn import naive_bayes
from sklearn import metrics
from sklearn import preprocessing
from sklearn import neighbors

## 1.	Read in the training dataset.

In [258]:
dat = pd.read_csv("C:/Users/Shalini/Desktop/data methods python/train_data.csv")
dat.head(10).transpose()

0       1       2        3        4       5  \
word_freq_make               0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_address            0.900   0.000   0.000    0.220    0.000   0.000   
word_freq_all                0.000   0.000   0.000    0.220    0.000   0.000   
word_freq_3d                 0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_our                0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_over               0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_remove             0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_internet           0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_order              0.000   0.000   0.000    0.220    0.000   0.000   
word_freq_mail               0.000   1.000   0.000    2.750    0.000   0.000   
word_freq_receive            0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_will               0.000   0.000   0.840    0.680    0.330   0.000   
word_freq_people             0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_report             0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_addresses          0.000   0.000   0.000    0.680    0.000   0.000   
word_freq_free               0.900   0.000   0.000    0.450    0.000   0.000   
word_freq_business           0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_email              1.810   0.000   0.000    1.370    0.000   3.570   
word_freq_you                3.630   2.000   5.040    2.060    0.000   0.000   
word_freq_credit             0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_your               2.720   0.000   0.840    4.120    0.000   0.000   
word_freq_font               0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_000                0.000   0.000   0.000    0.450    0.000   0.000   
word_freq_money              0.000   0.000   0.000    0.220    0.000   0.000   
word_freq_hp                 0.000   0.000   0.000    0.000    0.000   7.140   
word_freq_hpl                0.000   0.000   0.000    0.000    0.000   3.570   
word_freq_george             0.000   0.000   0.000    0.000    0.670   0.000   
word_freq_650                0.000   0.000   0.000    0.000    0.000   3.570   
word_freq_lab                0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_labs               0.000   0.000   0.000    0.000    0.000   3.570   
word_freq_telnet             0.000   0.000   0.000    0.000    2.710   3.570   
word_freq_857                0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_data               0.000   0.000   0.000    0.000    0.670   0.000   
word_freq_415                0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_85                 0.000   0.000   0.000    0.000    0.000   3.570   
word_freq_technology         0.000   0.000   0.000    0.220    0.000   3.570   
word_freq_1999               0.000   1.000   0.000    0.000    0.000   0.000   
word_freq_parts              0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_pm                 0.000   0.000   0.000    0.000    0.670   0.000   
word_freq_direct             0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_cs                 0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_meeting            0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_original           0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_project            0.000   0.000   0.000    0.000    4.400   0.000   
word_freq_re                 0.900   0.000   0.840    0.000    0.000   0.000   
word_freq_edu                0.000   5.000   0.000    0.000    0.000   0.000   
word_freq_table              0.000   0.000   0.000    0.000    0.000   0.000   
word_freq_conference         0.000   0.000   0.000    0.000    0.000   0.000   
char_freq_;                  0.000   0.000   0.000    0.028    0.000   0.000   
char_freq_(                  0.00

In [259]:
dat.dtypes

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge

In [260]:
dat.corr(method='pearson')['class']

word_freq_make                0.128547
word_freq_address            -0.028764
word_freq_all                 0.180645
word_freq_3d                  0.056442
word_freq_our                 0.247632
word_freq_over                0.262313
word_freq_remove              0.326272
word_freq_internet            0.214545
word_freq_order               0.222162
word_freq_mail                0.125602
word_freq_receive             0.220357
word_freq_will                0.019502
word_freq_people              0.145564
word_freq_report              0.071636
word_freq_addresses           0.199327
word_freq_free                0.281547
word_freq_business            0.274075
word_freq_email               0.198212
word_freq_you                 0.267893
word_freq_credit              0.178759
word_freq_your                0.381736
word_freq_font                0.092107
word_freq_000                 0.338968
word_freq_money               0.222476
word_freq_hp                 -0.250884
word_freq_hpl            

## 2. X_train

In [261]:
# Considering all the features except for target variable because a spam mail has words. The frequency of the words may or maynot
#be highly correlated with target variable but they have significance in finding if a e-mail is spam or not.
X_train = dat[['word_freq_make','word_freq_address','word_freq_all','word_freq_3d','word_freq_our',
                 'word_freq_over','word_freq_remove','word_freq_internet','word_freq_order','word_freq_mail',
                 'word_freq_receive','word_freq_will','word_freq_people','word_freq_report','word_freq_addresses',
                 'word_freq_free','word_freq_business','word_freq_email','word_freq_you','word_freq_credit','word_freq_your', 
                 'word_freq_font','word_freq_000','word_freq_money','word_freq_hpl','word_freq_george','word_freq_650',
                 'word_freq_lab','word_freq_labs','word_freq_telnet','word_freq_857','word_freq_415','word_freq_85',
                 'word_freq_technology','word_freq_1999','word_freq_parts','word_freq_pm','word_freq_direct','word_freq_cs',
                 'word_freq_meeting','word_freq_original','word_freq_project','word_freq_re','word_freq_edu',
                 'word_freq_table','word_freq_conference','char_freq_;','char_freq_(','char_freq_[','char_freq_!',
                  'char_freq_$','char_freq_#','capital_run_length_average','capital_run_length_longest','capital_run_length_total']]

In [262]:
X_train.describe().transpose()

count        mean         std  min   25%     50%  \
word_freq_make              3220.0    0.103264    0.303500  0.0   0.0   0.000   
word_freq_address           3220.0    0.212137    1.284390  0.0   0.0   0.000   
word_freq_all               3220.0    0.285643    0.510983  0.0   0.0   0.000   
word_freq_3d                3220.0    0.062012    1.334705  0.0   0.0   0.000   
word_freq_our               3220.0    0.312714    0.677586  0.0   0.0   0.000   
word_freq_over              3220.0    0.097652    0.259596  0.0   0.0   0.000   
word_freq_remove            3220.0    0.114876    0.399012  0.0   0.0   0.000   
word_freq_internet          3220.0    0.102121    0.375549  0.0   0.0   0.000   
word_freq_order             3220.0    0.083475    0.267773  0.0   0.0   0.000   
word_freq_mail              3220.0    0.238975    0.657438  0.0   0.0   0.000   
word_freq_receive           3220.0    0.059736    0.203723  0.0   0.0   0.000   
word_freq_will              3220.0    0.531273    0.835483  0.0   0.0   0.090   
word_freq_people            3220.0    0.087807    0.271794  0.0   0.0   0.000   
word_freq_report            3220.0    0.056829    0.300213  0.0   0.0   0.000   
word_freq_addresses         3220.0    0.046379    0.253439  0.0   0.0   0.000   
word_freq_free              3220.0    0.249537    0.803401  0.0   0.0   0.000   
word_freq_business          3220.0    0.143149    0.437734  0.0   0.0   0.000   
word_freq_email             3220.0    0.177435    0.514909  0.0   0.0   0.000   
word_freq_you               3220.0    1.665320    1.802579  0.0   0.0   1.310   
word_freq_credit            3220.0    0.086022    0.542223  0.0   0.0   0.000   
word_freq_your              3220.0    0.798022    1.181653  0.0   0.0   0.170   
word_freq_font              3220.0    0.107252    0.969864  0.0   0.0   0.000   
word_freq_000               3220.0    0.104453    0.352068  0.0   0.0   0.000   
word_freq_money             3220.0    0.090615    0.422020  0.0   0.0   0.000   
word_freq_hpl               3220.0    0.269786    0.941306  0.0   0.0   0.000   
word_freq_george            3220.0    0.773543    3.421963  0.0   0.0   0.000   
word_freq_650               3220.0    0.116326    0.497956  0.0   0.0   0.000   
word_freq_lab               3220.0    0.097730    0.596303  0.0   0.0   0.000   
word_freq_labs              3220.0    0.103075    0.466910  0.0   0.0   0.000   
word_freq_telnet            3220.0    0.062550    0.365604  0.0   0.0   0.000   
word_freq_857               3220.0    0.049590    0.346865  0.0   0.0   0.000   
word_freq_415               3220.0    0.049839    0.346899  0.0   0.0   0.000   
word_freq_85                3220.0    0.099935    0.443431  0.0   0.0   0.000   
word_freq_technology        3220.0    0.098410    0.398681  0.0   0.0   0.000   
word_freq_1999              3220.0    0.137196    0.425290  0.0   0.0   0.000   
word_freq_parts             3220.0    0.014624    0.234919  0.0   0.0   0.000   
word_freq_pm                3220.0    0.072938    0.401819  0.0   0.0   0.000   
word_freq_direct            3220.0    0.069475    0.371930  0.0   0.0   0.000   
word_freq_cs                3220.0    0.041786    0.336403  0.0   0.0   0.000   
word_freq_meeting           3220.0    0.133298    0.792959  0.0   0.0   0.000   
word_freq_original          3220.0    0.043752    0.213012  0.0   0.0   0.000   
word_freq_project           3220.0    0.068497    0.550429  0.0   0.0   0.000   
word_freq_re                3220.0    0.304019    1.011806  0.0   0.0   0.000   
word_freq_edu               3220.0    0.179484    0.923475  0.0   0.0   0.000   
word_freq_table             3220.0    0.005752    0.078274  0.0   0.0   0.000   
word_freq_conference        3220.0    0.030714    0.285833  0.0   0.0   0.000   
char_freq_;                 3220.0    0.036741    0.216996  0.0   0.0   0.000   
char_freq_(                 3220.0    0.141395    0.294814  0.0   0.0   0.065   
char_freq_[                 3220.0    0.016743    0.090718  

## 3. y_train with target variable 

In [263]:
y_train = dat['class'].copy()
y_train.head()

0    1
1    0
2    0
3    1
4    0
Name: class, dtype: int64

## 4.	Read in the test dataset.

In [264]:
dat1 = pd.read_csv("C:/Users/Shalini/Desktop/data methods python/test_data.csv")
dat1.head(10).transpose()

0      1       2       3      4       5  \
word_freq_make                 0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_address              0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_all                  0.150  0.000   0.000   4.340   0.00   0.590   
word_freq_3d                   0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_our                  0.300  0.000   1.290   0.000   0.00   0.590   
word_freq_over                 0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_remove               0.150  0.000   0.000   0.000   0.00   0.000   
word_freq_internet             0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_order                0.610  0.000   0.000   0.000   0.00   0.000   
word_freq_mail                 0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_receive              0.300  0.000   0.000   0.000   0.00   0.000   
word_freq_will                 0.000  0.000   0.000   0.000   1.24   2.380   
word_freq_people               0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_report               0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_addresses            0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_free                 0.920  0.000   0.640   0.000   0.00   0.000   
word_freq_business             0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_email                1.530  0.000   1.290   0.000   0.00   1.190   
word_freq_you                  0.610  0.000   2.580   0.000   0.00   0.590   
word_freq_credit               0.000  0.000   0.640   0.000   0.00   0.000   
word_freq_your                 0.000  0.000   1.290   0.000   0.00   0.000   
word_freq_font                 0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_000                  0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_money                0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_hp                   0.000  0.000   0.000   0.000   0.00   1.780   
word_freq_hpl                  0.000  0.000   0.000   0.000   0.00   1.190   
word_freq_george               0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_650                  0.000  0.000   0.000   0.000   0.00   0.590   
word_freq_lab                  0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_labs                 0.000  0.000   0.000   0.000   0.00   0.590   
word_freq_telnet               0.000  0.000   0.000   0.000   0.00   0.590   
word_freq_857                  0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_data                 0.000  0.000   0.000   0.000   6.22   0.000   
word_freq_415                  0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_85                   0.000  0.000   0.000   0.000   0.00   0.590   
word_freq_technology           0.000  0.000   0.000   0.000   0.00   0.590   
word_freq_1999                 0.300  0.000   0.000   0.000   0.00   0.000   
word_freq_parts                0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_pm                   0.000  0.000   0.640   0.000   0.00   1.190   
word_freq_direct               0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_cs                   0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_meeting              0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_original             0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_project              0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_re                   0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_edu                  0.000  9.090   0.000   0.000   0.00   0.000   
word_freq_table                0.000  0.000   0.000   0.000   0.00   0.000   
word_freq_conference           0.000  0.000   0.000   0.000   0.00   0.000   
char_freq_;                    0.000  0.000   0.000   0.000   0.00   0.000   
char_freq_(                    0.083  0.719   0.224   0.000   0.00   0.177   
char_freq_[                    0.000  0.000   0.000   0.0

## 5.	X_test

In [265]:
# Concat all dummies and continuous variables to create predictors X
X_test = dat1[['word_freq_make','word_freq_address','word_freq_all','word_freq_3d','word_freq_our',
                 'word_freq_over','word_freq_remove','word_freq_internet','word_freq_order','word_freq_mail',
                 'word_freq_receive','word_freq_will','word_freq_people','word_freq_report','word_freq_addresses',
                 'word_freq_free','word_freq_business','word_freq_email','word_freq_you','word_freq_credit','word_freq_your', 
                 'word_freq_font','word_freq_000','word_freq_money','word_freq_hpl','word_freq_george','word_freq_650',
                 'word_freq_lab','word_freq_labs','word_freq_telnet','word_freq_857','word_freq_415','word_freq_85',
                 'word_freq_technology','word_freq_1999','word_freq_parts','word_freq_pm','word_freq_direct','word_freq_cs',
                 'word_freq_meeting','word_freq_original','word_freq_project','word_freq_re','word_freq_edu',
                 'word_freq_table','word_freq_conference','char_freq_;','char_freq_(','char_freq_[','char_freq_!',
                  'char_freq_$','char_freq_#','capital_run_length_average','capital_run_length_longest','capital_run_length_total']]

In [266]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1381 entries, 0 to 1380
Data columns (total 55 columns):
word_freq_make                1381 non-null float64
word_freq_address             1381 non-null float64
word_freq_all                 1381 non-null float64
word_freq_3d                  1381 non-null float64
word_freq_our                 1381 non-null float64
word_freq_over                1381 non-null float64
word_freq_remove              1381 non-null float64
word_freq_internet            1381 non-null float64
word_freq_order               1381 non-null float64
word_freq_mail                1381 non-null float64
word_freq_receive             1381 non-null float64
word_freq_will                1381 non-null float64
word_freq_people              1381 non-null float64
word_freq_report              1381 non-null float64
word_freq_addresses           1381 non-null float64
word_freq_free                1381 non-null float64
word_freq_business            1381 non-null float64
word_freq

## 6.	 y_test

In [267]:
y_test= dat1['class'].copy()
y_test.head()

0    1
1    0
2    1
3    0
4    0
Name: class, dtype: int64

In [268]:
y_test.describe()

count    1381.000000
mean        0.393917
std         0.488794
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: class, dtype: float64

## 7.1. K-Nearest Neighbors (kNN)

### 7.2. Train a k-NN Classifier

In [269]:
from sklearn import neighbors
import numpy as np

In [270]:
# Candidate values of k
k = np.arange(50) + 1

k

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])

In [271]:
para_knn = {'n_neighbors':k}
para_knn

{'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
        35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])}

In [272]:
knn = neighbors.KNeighborsClassifier()

In [273]:
grid_knn = model_selection.GridSearchCV(knn, para_knn)

In [274]:
grid_knn.fit(X_train,y_train.values.ravel())

GridSearchCV(cv=None, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [279]:
# Show best parameters
grid_knn.best_params_


{'n_neighbors': 1}

In [281]:
# KNN: K=5, default measure of distance (euclidean)
knn1 = neighbors.KNeighborsClassifier(n_neighbors=1, 
                                      weights='uniform', 
                                      algorithm='auto')# auto euclidian theorem!

In [284]:
knn1.fit(X_train, y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

### 7.3 Evaluate k-NN Classifier

In [285]:
pred_y = knn1.predict(X_test)

In [286]:
# Print confusion matrix
print(metrics.confusion_matrix(y_test, pred_y))

[[718 119]
 [141 403]]


In [287]:
# Calculate classification accuracy
metrics.accuracy_score(y_test, pred_y)

0.8117306299782766

In [288]:
# Calculate AUC
metrics.roc_auc_score(y_test, pred_y)# good classifier AUC>0.8

0.7993171955161994

In [289]:
# Calculate Cohen's Kappa
metrics.cohen_kappa_score(y_test, pred_y)

0.6029019807345639

In [290]:
# Print classification report
print(metrics.classification_report(y_test, pred_y))

             precision    recall  f1-score   support

          0       0.84      0.86      0.85       837
          1       0.77      0.74      0.76       544

avg / total       0.81      0.81      0.81      1381



## 8.1. Decision Tree

### 8.2 Train a Classification Tree

In [205]:
from sklearn import tree

In [206]:
np.arange(2,51)

array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])

In [207]:
para_dt = {'min_samples_split':np.arange(2,51)}
para_dt

{'min_samples_split': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
        19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
        36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])}

In [208]:
dsnt = tree.DecisionTreeClassifier()

In [210]:
grid_dsnt = model_selection.GridSearchCV(dsnt, para_dt)
grid_dsnt.fit(X_train,y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'min_samples_split': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [212]:
# Show best parameters
grid_dsnt.best_params_


{'min_samples_split': 30}

### 8.3 Evaluate Classification Tree

In [214]:
# Predict on test set
pred_y30 = grid_dsnt.predict(X_test)

In [215]:
# Print confusion matrix
print(metrics.confusion_matrix(y_test, pred_y30))

[[769  68]
 [ 76 468]]


In [217]:
# Calculate classification accuracy
metrics.accuracy_score(y_test, pred_y30)

0.8957277335264301

In [218]:
# Calculate AUC
metrics.roc_auc_score(y_test, pred_y30)

0.8895257923958114

In [219]:
# Calculate Cohen's Kappa
metrics.cohen_kappa_score(y_test, pred_y30)

0.7810620139335382

In [220]:
# Print classification report
print(metrics.classification_report(y_test, pred_y30))

             precision    recall  f1-score   support

          0       0.91      0.92      0.91       837
          1       0.87      0.86      0.87       544

avg / total       0.90      0.90      0.90      1381



## 9.1. Artificial Neural Network (MLP)

### 9.2. Train an MLP

In [221]:
from sklearn import neural_network

In [222]:
para_ann = {'hidden_layer_sizes':[(300,), (250,), (200,), 
                     (150,),  (100,), (50,), (50,10), (9,7,5,3)]}
para_ann

{'hidden_layer_sizes': [(300,),
  (250,),
  (200,),
  (150,),
  (100,),
  (50,),
  (50, 10),
  (9, 7, 5, 3)]}

In [223]:
ann = neural_network.MLPClassifier(max_iter=500)

In [224]:
grid_ann = model_selection.GridSearchCV(ann, para_ann)

In [225]:
grid_ann.fit(X_train,y_train.values.ravel())

GridSearchCV(cv=None, error_score='raise',
       estimator=MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'hidden_layer_sizes': [(300,), (250,), (200,), (150,), (100,), (50,), (50, 10), (9, 7, 5, 3)]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [226]:
# Show best parameters
grid_ann.best_params_

{'hidden_layer_sizes': (50,)}

### 9.3. Evaluate the MLP

In [227]:
# Predict on test set
pred_yhd = grid_ann.predict(X_test)

In [228]:
# Print confusion matrix
print(metrics.confusion_matrix(y_test, pred_yhd))

[[769  68]
 [ 54 490]]


In [231]:
# Calculate classification accuracy
metrics.accuracy_score(y_test, pred_yhd)

0.9116582186821144

In [232]:
# Calculate AUC
metrics.roc_auc_score(y_test, pred_yhd)

0.9097463806311056

In [233]:
# Calculate Cohen's Kappa
metrics.cohen_kappa_score(y_test, pred_yhd)

0.8158179540381172

In [234]:
# Print classification report
print(metrics.classification_report(y_test, pred_yhd))

             precision    recall  f1-score   support

          0       0.93      0.92      0.93       837
          1       0.88      0.90      0.89       544

avg / total       0.91      0.91      0.91      1381



## 10.1 Random Forests

### 10.2 Train a Random Forests

In [235]:
from sklearn import ensemble

In [237]:
rf = ensemble.RandomForestClassifier()

In [236]:
para_rf = [{'n_estimators':[3, 10, 20, 30],'max_features':[2,4,6,8,10]},
           {'bootstrap':[False],'n_estimators':[3,10,20],'max_features':[2,3,4]}]
para_rf

[{'max_features': [2, 4, 6, 8, 10], 'n_estimators': [3, 10, 20, 30]},
 {'bootstrap': [False],
  'max_features': [2, 3, 4],
  'n_estimators': [3, 10, 20]}]

In [238]:
grid_rf = model_selection.GridSearchCV(rf, para_rf)

In [239]:
grid_rf.fit(X_train,y_train.values.ravel())

GridSearchCV(cv=None, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid=[{'n_estimators': [3, 10, 20, 30], 'max_features': [2, 4, 6, 8, 10]}, {'bootstrap': [False], 'n_estimators': [3, 10, 20], 'max_features': [2, 3, 4]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [240]:
# Show best parameters
grid_rf.best_params_


{'bootstrap': False, 'max_features': 3, 'n_estimators': 20}

### 10.3. Evaluate the Random Forests

In [241]:
# Predict on test set
pred_y_rf = grid_rf.predict(X_test)

In [242]:
# Print confusion matrix
print(metrics.confusion_matrix(y_test, pred_y_rf))

[[813  24]
 [ 42 502]]


In [243]:
# Calculate classification accuracy
metrics.accuracy_score(y_test, pred_y_rf)

0.9522085445329471

In [245]:
# Calculate AUC
metrics.roc_auc_score(y_test, pred_y_rf)

0.9470601412608054

In [247]:
# Calculate Cohen's Kappa
metrics.cohen_kappa_score(y_test, pred_y_rf)

0.8993286811533695

In [248]:
# Print classification report
print(metrics.classification_report(y_test, pred_y_rf))

             precision    recall  f1-score   support

          0       0.95      0.97      0.96       837
          1       0.95      0.92      0.94       544

avg / total       0.95      0.95      0.95      1381



### 11. Comparing Models

In [295]:
para = [grid_knn.best_params_,grid_dsnt.best_params_,grid_ann.best_params_, grid_rf.best_params_]

acc = [metrics.accuracy_score(y_test, pred_y),
       metrics.accuracy_score(y_test, pred_y30),
       metrics.accuracy_score(y_test, pred_yhd),
       metrics.accuracy_score(y_test, pred_y_rf)]

pre = [metrics.precision_score(y_test, pred_y),
       metrics.precision_score(y_test, pred_y30),
       metrics.precision_score(y_test, pred_yhd),
       metrics.precision_score(y_test, pred_y_rf)]

rec = [metrics.recall_score(y_test, pred_y),
       metrics.recall_score(y_test, pred_y30),
       metrics.recall_score(y_test, pred_yhd),
       metrics.recall_score(y_test, pred_y_rf)]

f1  = [metrics.f1_score(y_test, pred_y),
       metrics.f1_score(y_test, pred_y30),
       metrics.f1_score(y_test, pred_yhd),
       metrics.f1_score(y_test, pred_y_rf)]

In [296]:
pd.DataFrame({'Parameters':para,
              'Accuracy':acc, 'Precision':pre, 'Recall':rec,'F1 Score': f1},
             index = ['k-NN','Classification Tree','Neural Network', 'Random Forests'])

Accuracy  F1 Score  \
k-NN                 0.811731  0.756098   
Classification Tree  0.895728  0.866667   
Neural Network       0.911658  0.889292   
Random Forests       0.952209  0.938318   

                                                            Parameters  \
k-NN                                                {'n_neighbors': 1}   
Classification Tree                          {'min_samples_split': 30}   
Neural Network                           {'hidden_layer_sizes': (50,)}   
Random Forests       {'bootstrap': False, 'max_features': 3, 'n_est...   

                     Precision    Recall  
k-NN                  0.772031  0.740809  
Classification Tree   0.873134  0.860294  
Neural Network        0.878136  0.900735  
Random Forests        0.954373  0.922794

I can see that the f1 score for the model is highest for Random forests. Thus, I would like to choos Random forests model with 
F1 score 0.938 to validate if the email is spam or not